autogpt/utils/agent_discovery.py

```
@@ -0,0 +1,111 @@
+        DiscoveryResult if the domain publishes agent services, None otherwise.
+    """
+    import time
+
+    if use_cache and domain in _cache:
+        cached_at, cached_result = _cache[domain]
+        if time.time() - cached_at < _CACHE_TTL:
+            return DiscoveryResult(cached_result) if cached_result else None
+
+    url = f"https://{domain}/.well-known/agent-discovery.json"
+    try:
+        req = urllib.request.Request(url, headers={"User-Agent": "agent-discovery/0.1"})
```

Ruff (0.15.9)

[error] Audit URL open for permitted schemes. Allowing use of `file:` or custom schemes is often unexpected.

(S310)